doc/Dump-Utility.md

```diff
@@ -0,0 +1,403 @@
+
+## About this Manual
+This document describes the details of a dump cli utility which collects and dumps the redis state for a given feature/module. 
+
+## Definitions/Abbreviations 
+###### Table 1: Abbreviations
+| Abbreviation | Description                                                  |
+| ------------ | ------------------------------------------------------------ |
+| APP DB       | Application Database                                         |
+| ASIC DB      | ASIC Database                                                |
+| CONF DB      | Configuration Database                                       |
+
+
+                     
+In this document, the term '**redis state**' refers to the intermediate state of a given feature present across all the Redis DB's 
+
+## Overview 
+In SONiC, there usually exists a set of tables related/relevant to a particular module. All of these have to be looked at to confirm whether  any configuration update is properly applied and propagated.
+
+The task of debugging quickly becomes tedious because currently, there is no utility which does print a unified view of the redis-state.
+This is the problem which is addressed by this dump utility.
+
+This utility provides the base infrastructure and guidelines to make is easy for the developers to extend and support the utility for different modules.
+
+## **1 Requirements**
+
+### 1.1 Intended Usage
+
+Note: Currently, `state` is the only subcommand under `dump`.
+
+```
+dump state <feature/module> <arg>
+
+Example usage for Port feature which accepts an argument port_name
+dump state port <port_name>
+dump state port Ethernet4
+dump state port Ethernet4,Ethernet8,Ethernet12
+dump state port all
+```
+
+#### 1.1 Intended Usage
+1) The `dump state` subcommand will take a feature/module or any logical top-level name as the first argument. 
+2) A Second Argument should also be defined for a feature. 
+3) This argument could either be a table-key or a unique-field-value present in either Conf DB or Appl DB.
+   * Eg: For PORT, the second argument will be an interface name i.e 'Ethernet128' which is a table-key. On the other hand, the secondary argument for COPP will be a trap_id such as  'arp_req', which is a field-value and not a key of any table.
+4) The decision of what to pass as a secondary argument lies with the discretion of the one who is writing the module.
+6) The Command should also take a list of comma seperated inputs for the secondary argument
+7) The Command should also accept an "all" value and which means it should print the unified view for every entry related to that feature.
+
+```
+root@sonic# dump state --help
+Usage: dump state [OPTIONS] MODULE ARG
+
+  Dump the redis state of the module specified
+
+Options:
+  -s, --show     Display all the Modules and the name of their correponding secondary argument 
+  -d, --db TEXT  Only dump from these Databases
+  -t, --table    Print in tabular format  [default: False]
```

Comment:
@vivekreddynv please provide some CLI examples for general usage and for different options, e.g.:
```
root@sonic# dump state copp Ethernet0
...

root@sonic# dump state copp Ethernet0 --table
...

root@sonic# dump state copp Ethernet0 --rid
...

root@sonic# dump state copp Ethernet0 --key-map
...

root@sonic# dump state copp Ethernet0 --verbose
...
```

Reply:
Done!, Thanks for the input

Reply:
Only, the tabular format is left.

Reply:
Tabular format output completed